### PR TITLE
feat(schematics): table schematic

### PR DESCRIPTION
**Pull request by amcdnl**
This PR creates a table schematic.

Source: https://github.com/amcdnl/material-example-schematics/tree/schematic-table

### CODE REVIEW

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
+import { Component, OnInit, ViewChild, <% if(!!viewEncapsulation) { %>, ViewEncapsulation<% }%><% if(changeDetection !== 'Default') { %>, ChangeDetectionStrategy<% }%> } from '@angular/core';
+import { Observable } from 'rxjs/Observable';
+import { map } from 'rxjs/operators/map';
+import { MatPaginator, MatSort } from '@angular/material';
+import { DataSource } from '@angular/cdk/collections';
+import { of } from 'rxjs/observable/of';
+import { merge } from 'rxjs/observable/merge';
+
+export interface Element {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: `Element` is already a thing. I'd just make this something like `DataItem` with a comment (a TODO?) of something like "Replace this with your own data model type"

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  name: string;
+  position: number;
+  weight: number;
+  symbol: string;
+}
+
+export class ExampleDataSource extends DataSource<any> {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: The data source should go in its own file. I'd like this data source to play a big role in guiding users in structuring their data access and manipulation.

It would be great if we could add an option for a client data-source vs. a server data source. The client one would perform everything locally and synchronously, while the server one would call out to (fake) server methods and do thing asynchronously.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    ];
+
+    return merge(...stream).pipe(map(() => {
+      return this.getPagedData(this.getSortedData([...this.data]));
+    }));
+  }
+
+  disconnect() {}
+
+  getPagedData(data) {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Make `getPagedData` and `getSortedData` private?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    {position: 19, name: 'Potassium', weight: 39.0983, symbol: 'K'},
+    {position: 20, name: 'Calcium', weight: 40.078, symbol: 'Ca'},
+  ];
+
+  constructor(private _paginator: MatPaginator, private _sort: MatSort) {
+    super();
+  }
+
+  connect(): Observable<Element[]> {
+    const stream = [
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: I'd rename `stream` to something like `dataMutations` or `dataUpdates`, and add a comment like
```ts
// Combine everything that affects the rendered data into one update
// stream for the data-table to consume.
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 5]
```diff
@@ -0,0 +1,132 @@
+import { Component, OnInit, ViewChild, <% if(!!viewEncapsulation) { %>, ViewEncapsulation<% }%><% if(changeDetection !== 'Default') { %>, ChangeDetectionStrategy<% }%> } from '@angular/core';
+import { Observable } from 'rxjs/Observable';
+import { map } from 'rxjs/operators/map';
+import { MatPaginator, MatSort } from '@angular/material';
+import { DataSource } from '@angular/cdk/collections';
+import { of } from 'rxjs/observable/of';
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: `observableOf`

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+  connect(): Observable<Element[]> {
+    const stream = [
+      of(this.data),
+      this._paginator.page,
+      this._sort.sortChange
+    ];
+
+    return merge(...stream).pipe(map(() => {
+      return this.getPagedData(this.getSortedData([...this.data]));
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: `[...this.data]` is equivalent to `this.data`

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+  getPagedData(data) {
+    const startIndex = this._paginator.pageIndex * this._paginator.pageSize;
+    return data.splice(startIndex, this._paginator.pageSize);
+  }
+
+  getSortedData(data) {
+    if (!this._sort.active || this._sort.direction === '') { return data; }
+    return data.sort((a, b) => {
+      const propertyA = a[this._sort.active];
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Referencing a property via string isn't compatible with minifiers that do property renaming (e.g. Closure Compiler).

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+      const propertyB = b[this._sort.active];
+      const valueA = isNaN(+propertyA) ? propertyA : +propertyA;
+      const valueB = isNaN(+propertyB) ? propertyB : +propertyB;
+      return (valueA < valueB ? -1 : 1) * (this._sort.direction === 'asc' ? 1 : -1);
+    });
+  }
+}
+
+@Component({
+  selector: '<%= selector %>',<% if(inlineTemplate) { %>
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: @hansl are there any plans to make it easier to do templateUrl vs inline template without duplicating?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        [length]="20"
+        [pageIndex]="0"
+        [pageSize]="10"
+        [pageSizeOptions]="[5, 10, 25, 100]">
+      </mat-paginator>
+    </div>
+  `,<% } else { %>
+  templateUrl: './<%= dasherize(name) %>.component.html',<% } if(inlineStyle) { %>
+  styles: []<% } else { %>
+  styleUrls: ['./<%= dasherize(name) %>.component.<%= styleext %>']<% } %><% if(!!viewEncapsulation) { %>,
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Is the `!!` necessary?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  encapsulation: ViewEncapsulation.<%= viewEncapsulation %><% } if (changeDetection !== 'Default') { %>,
+  changeDetection: ChangeDetectionStrategy.<%= changeDetection %><% } %>
+})
+export class <%= classify(name) %>Component implements OnInit {
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+  @ViewChild(MatSort) sort: MatSort;
+  displayedColumns = ['position', 'name', 'weight', 'symbol'];
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Add a description to this like
```ts
/** Columns displayed in the table. Columns IDs can be added, removed, or reordered. */
```

File: schematics/table/index.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  return chain([
+    buildComponent({ ...options }),
+    options.skipImport ? noop() : addNavModulesToModule(options)
+  ]);
+}
+
+/**
+ * Adds the required modules to the relative module.
+ */
+function addNavModulesToModule(options: Schema) {
```
Comment (jelbourn) [inline schematics/table/index.ts]: Nav Modules?

File: schematics/table/index.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+import {findModuleFromOptions} from '../utils/devkit-utils/find-module';
+import {buildComponent} from '../utils/devkit-utils/component';
+
+/**
+ * Scaffolds a new table component.
+ * Internally it bootstraps the base component schematic
+ */
+export default function(options: Schema): Rule {
+  return chain([
+    buildComponent({ ...options }),
```
Comment (jelbourn) [inline schematics/table/index.ts]: Omit spaces in braces.

File: schematics/table/index_spec.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+
+  beforeEach(() => {
+    runner = new SchematicTestRunner('schematics', collectionPath);
+  });
+
+  it('should create table files and add them to module', () => {
+    const tree = runner.runSchematic('materialTable', { ...options }, createTestApp());
+    const files = tree.files;
+
+    expect(files.indexOf('/src/app/foo/foo.component.css')).toBeGreaterThanOrEqual(0);
```
Comment (jelbourn) [inline schematics/table/index_spec.ts]: You should be able to use the `toContain` matcher:
```ts
expect(files).toContain('/src/app/foo/foo.component.css');
...
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 9 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: I didn't write this, I copied the code from the default component template for this section.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  beforeEach(() => {
+    fixture = TestBed.createComponent(<%= classify(name) %>Component);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+});
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts]: Missing a newline at the end of the file.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
+
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { <%= classify(name) %>Component } from './<%= dasherize(name) %>.component';
+
+describe('<%= classify(name) %>Component', () => {
+  let component: <%= classify(name) %>Component;
+  let fixture: ComponentFixture<<%= classify(name) %>Component>;
+
+  beforeEach(async(() => {
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts]: Maybe this should be `fakeAsync`, instead of `async`? We've moved over most of our components to be `fakeAsync`.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(<%= classify(name) %>Component);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts]: The test name sounds off. Maybe something along the lines of `should compile`?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+  let fixture: ComponentFixture<<%= classify(name) %>Component>;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [ <%= classify(name) %>Component ]
+    })
+    .compileComponents();
+  }));
+
+  beforeEach(() => {
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts]: You can combine these two `beforeEach` calls into one.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 18]
```diff
@@ -0,0 +1,132 @@
+import { Component, OnInit, ViewChild, <% if(!!viewEncapsulation) { %>, ViewEncapsulation<% }%><% if(changeDetection !== 'Default') { %>, ChangeDetectionStrategy<% }%> } from '@angular/core';
+import { Observable } from 'rxjs/Observable';
+import { map } from 'rxjs/operators/map';
+import { MatPaginator, MatSort } from '@angular/material';
+import { DataSource } from '@angular/cdk/collections';
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Should we have the spaces around the curly braces here? We don't have them in our source.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+      this._paginator.page,
+      this._sort.sortChange
+    ];
+
+    return merge(...stream).pipe(map(() => {
+      return this.getPagedData(this.getSortedData([...this.data]));
+    }));
+  }
+
+  disconnect() {}
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Maybe there should be an example of what cleaning up here would look like, or a comment that says something like "Don't forget to clean up!".

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+
+  getPagedData(data) {
+    const startIndex = this._paginator.pageIndex * this._paginator.pageSize;
+    return data.splice(startIndex, this._paginator.pageSize);
+  }
+
+  getSortedData(data) {
+    if (!this._sort.active || this._sort.direction === '') { return data; }
+    return data.sort((a, b) => {
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: nit: missing an extra space above the `return`. Also switching the `if` to multiple lines.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 21]
```diff
@@ -0,0 +1,132 @@
...
+      const propertyA = a[this._sort.active];
+      const propertyB = b[this._sort.active];
+      const valueA = isNaN(+propertyA) ? propertyA : +propertyA;
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Since these are also supposed to serve as an example, it might be more appropriate to use something more readable like `parseFloat(property) || 0`.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 22]
```diff
@@ -0,0 +1,132 @@
...
+          <mat-cell *matCellDef="let element">{{element.name}}</mat-cell>
+        </ng-container>
+
+        <!-- Weight Column -->
+        <ng-container matColumnDef="weight">
+          <mat-header-cell *matHeaderCellDef mat-sort-header>Weight</mat-header-cell>
+          <mat-cell *matCellDef="let element">{{element.weight}}</mat-cell>
+        </ng-container>
+
+        <!-- Color Column -->
```
Comment (crisbeto) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: The column is actually "symbol".

File: schematics/table/schema.json [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+      "default": "css"
+    },
+    "spec": {
+      "type": "boolean",
+      "description": "Specifies if a spec file is generated.",
+      "default": true
+    },
+    "flat": {
+      "type": "boolean",
+      "description": "Flag to indicate if a dir is created.",
```
Comment (crisbeto) [inline schematics/table/schema.json]: I don't think it's clear what a `dir` means in this case.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.html [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+
+    <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
+    <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
+  </mat-table>
+
+  <mat-paginator #paginator
+    [length]="20"
+    [pageIndex]="0"
+    [pageSize]="10"
+    [pageSizeOptions]="[5, 10, 25, 100]">
```
[the remark this answers is not in the document]
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.html]: Do we really want to encourage sometime to load 250 into view?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts (hunk 15 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts]: This was copied from the base component schematic.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts (hunk 16 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts]: This was copied from the base component schematic.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts (hunk 17 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.spec.ts]: This was copied from the base component schematic.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 18 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: The default schematics have them so I'd tend to stay the same for consistency. @jelbourn - what do u think?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 21 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Refactored to different implementation.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 22 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Changed to different property.

File: schematics/table/schema.json (hunk 23 above)
Comment (amcdnl) [reply schematics/table/schema.json]: This was copied from the base component schematic.

**Comment by amcdnl** [comment]
@jelbourn - Addressed all feedback minus the datasource. Do you wanna try to do that in this cut?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 9 above)
Comment (jelbourn) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Doesn't mean it shouldn't be corrected.

File: schematics/table/schema.json (hunk 23 above)
Comment (jelbourn) [reply schematics/table/schema.json]: Doesn't mean we can't make it better for ourselves

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.html [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+    </ng-container>
+
+    <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
+    <mat-row *matRowDef="let row; columns: displayedColumns;"></mat-row>
+  </mat-table>
+
+  <mat-paginator #paginator
+    [length]="dataSource.data.length"
+    [pageIndex]="0"
+    [pageSize]="10"
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.html]: Compromise on 50? 10 is just too small to ever be realistic.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 18 above)
Comment (jelbourn) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Yeah, we should be consistent with the rest of the CLI even if spaces in braces is nonsense.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 5 above)
Comment (jelbourn) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]:
```ts
import {of as observableOf} from 'rxjs/observable/of';
```
Since `of` as a function is confusing with `for of`

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 26]
```diff
@@ -0,0 +1,131 @@
+import { Component, OnInit, ViewChild, <% if(!!viewEncapsulation) { %>, ViewEncapsulation<% }%><% if(changeDetection !== 'Default') { %>, ChangeDetectionStrategy<% }%> } from '@angular/core';
+import { Observable } from 'rxjs/Observable';
+import { map } from 'rxjs/operators/map';
+import { MatPaginator, MatSort } from '@angular/material';
+import { DataSource } from '@angular/cdk/collections';
+import { of } from 'rxjs/observable/of';
+import { merge } from 'rxjs/observable/merge';
+
+/** Todo: Replace this with your own data model type */
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: `TODO` (all caps)

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+export interface DataItem {
+  name: string;
+  id: number;
+}
+
+function compare(a, b, isAsc) {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Can this name be more specific? `sortComparator`? Should also have a comment like 
```ts
/** Simple sort comparator for example ID/Name columns. */
```

I would also move this to the bottom of the file.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+  return (a < b ? -1 : 1) * (isAsc ? 1 : -1);
+}
+
+export class ExampleDataSource extends DataSource<any> {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Instead of "Example", could we use the name of the page that's being generated?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+  data: DataItem[] = [
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: I would move this array to be a constant at the top like `EXAMPLE_DATA`, and then when it's assigned in the class it should have a comment like
```
// TODO: replace this with real data from your application
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+    {id: 18, name: 'Argon'},
+    {id: 19, name: 'Potassium'},
+    {id: 20, name: 'Calcium'},
+  ];
+
+  constructor(private _paginator: MatPaginator, private _sort: MatSort) {
+    super();
+  }
+
+  connect(): Observable<Element[]> {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Add a method description like
```ts
/** 
 * Connect this data source to the table. The table will only update when
 * the returned stream emits new items.
 * @returns A stream of the items to be rendered.
 */
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+      this._paginator.page,
+      this._sort.sortChange
+    ];
+
+    return merge(...dataMutations).pipe(map(() => {
+      return this.getPagedData(this.getSortedData(this.data));
+    }));
+  }
+
+  disconnect() {}
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Add a comment inside the empty `disconnect` like
```ts
// TODO: clean up any open connections, free any held resources, etc.
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+
+  private getPagedData(data) {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Add method description

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+    const startIndex = this._paginator.pageIndex * this._paginator.pageSize;
+    return data.splice(startIndex, this._paginator.pageSize);
+  }
+
+  private getSortedData(data) {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Add method description

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 30 above)
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Return value should be `Observable<DataItem>`

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 8 above)
Comment (hansl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: It could be done manually, ie. put the template content in a variable that’s used both here and in the separate file. Read that value from a file on disk. There’s no silver bullet for this one.

Comment (hansl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: It could be done manually, ie. put the template content in a variable that’s used both here and in the separate file. Read that value from a file on disk. There’s no silver bullet for this one.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+    {id: 14, name: 'Silicon'},
+    {id: 15, name: 'Phosphorus'},
+    {id: 16, name: 'Sulfur'},
+    {id: 17, name: 'Chlorine'},
+    {id: 18, name: 'Argon'},
+    {id: 19, name: 'Potassium'},
+    {id: 20, name: 'Calcium'},
+  ];
+
+  constructor(private _paginator: MatPaginator, private _sort: MatSort) {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: I've been thinking about this, and I'm thinking it would be good to decouple the data source and the sorter/paginator. This would look something like the data source having its own sort and pagination state, and then using the outputs on the directives to update the data source state. 

My main reason for wanting to do this is to not make users think they *must* use those components.

File: schematics/table/schema.json (hunk 23 above)
Comment (amcdnl) [reply schematics/table/schema.json]: I agree but I wanted to keep this as close to the original as possible so when they support JSON schema inheritance, we can just do away with this file completely.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 30 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: I did one better and did: `<%= classify(name) %>Item[]`

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts (hunk 34 above)
Comment (amcdnl) [reply schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.ts]: Ya, I can see that but it wouldn't be near as clean then. What about a comment instead? Best way I could think of would be to:

```
this.dataSource = new DataSource({
  pageIndex: this.paginator.pageIndex,
  pageSize: this.paginator.pageSize
}, {
  active: this.sort.active,
  direction: this.sort.direction
});
```

**Comment by amcdnl** [comment]
@jelbourn - Addressed feedback.

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 35]
```diff
@@ -0,0 +1,100 @@
+import { merge } from 'rxjs/observable/merge';
+import { of as observableOf } from 'rxjs/observable/of';
+import { DataSource } from '@angular/cdk/collections';
+import { MatPaginator, MatSort } from '@angular/material';
+import { map } from 'rxjs/operators/map';
+import { Observable } from 'rxjs/Observable';
+
+/** TODO: Replace this with your own data model type */
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]: Switch to `//` style comments to be consistent with other TODOs

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+    }));
+  }
+
+  disconnect() {
+    // TODO: clean up any open connections, free any held resources, etc.
+  }
+
+  /**
+   * Client-side page the data by slicing out the next from the data array.
+   * If you are using external datasource for pagination, you would connect it here.
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]: I'd rephrase this as:
```
Paginate the data (client-side). If you're using server-side pagination,
this would be replaced by requesting the appropriate data from the server.
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 37]
```diff
@@ -0,0 +1,100 @@
...
+   */
+  private getPagedData(data: <%= classify(name) %>Item[]) {
+    const startIndex = this._paginator.pageIndex * this._paginator.pageSize;
+    return data.splice(startIndex, this._paginator.pageSize);
+  }
+
+  /**
+   * Client-side sort the data array.
+   * If you are using a external datasource for sorting, you would connect it here
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]: I'd rephrase this as:
```
Sort the data (client-side). If you're using server-side sorting,
this would be replaced by requesting the appropriate data from the server.
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 38]
```diff
@@ -0,0 +1,100 @@
...
+      switch (this._sort.active) {
+        case 'name': return compare(a.name, b.name, isAsc);
+        case 'id': return compare(+a.id, +b.id, isAsc);
+        default: return 0;
+      }
+    });
+  }
+}
+
+/** Simple sort comparator for example ID/Name columns. */
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]:
```ts
/** Simple sort comparator for example ID/Name columns (for client-side sorting). */
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 39]
```diff
@@ -0,0 +1,100 @@
...
+  {id: 14, name: 'Silicon'},
+  {id: 15, name: 'Phosphorus'},
+  {id: 16, name: 'Sulfur'},
+  {id: 17, name: 'Chlorine'},
+  {id: 18, name: 'Argon'},
+  {id: 19, name: 'Potassium'},
+  {id: 20, name: 'Calcium'},
+];
+
+export class <%= classify(name) %>DataSource extends DataSource<<%= classify(name) %>Item> {
```
Comment (jelbourn) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]: Add a class description like
```
Data source for the xxx view. This class should encapsulate all logic for fetching and
manipulating the displayed data (including sorting, pagination, and filtering).
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 40]
```diff
@@ -0,0 +1,100 @@
...
+  data: <%= classify(name) %>Item[] = EXAMPLE_DATA;
+
+  constructor(private _paginator: MatPaginator, private _sort: MatSort) {
```
Comment (andrewseguin) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]: Though we use underscores for private variables in Material, I believe in this case we should follow the standard Angular-way: 

> Avoid prefixing private properties and methods with an underscore.
https://angular.io/guide/styleguide#style-03-04

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 41]
```diff
@@ -0,0 +1,100 @@
...
+    super();
+  }
+
+  /**
+   * Connect this data source to the table. The table will only update when
+   * the returned stream emits new items.
+   * @returns A stream of the items to be rendered.
```
Comment (andrewseguin) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]: Do other schematics use jsdoc format (e.g. `@returns`)?

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts [hunk 42]
```diff
@@ -0,0 +1,100 @@
...
+      this._sort.sortChange
+    ];
+
+    return merge(...dataMutations).pipe(map(() => {
+      return this.getPagedData(this.getSortedData(this.data));
+    }));
+  }
+
+  disconnect() {
+    // TODO: clean up any open connections, free any held resources, etc.
```
Comment (andrewseguin) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__-datasource.ts]: This TODO would make me believe I _must_ do something here. Rather than TODO, maybe leave some jsdoc for the function.

```
/**
 * Called when the table is being destroyed. Use this function to clean up any open connections or free
 * any held resources that were set up during connect.
 */
disconnect() { }
```

File: schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.html [hunk 43]
```diff
@@ -0,0 +1,26 @@
+<div class="mat-elevation-z8">
+  <mat-table #table [dataSource]="dataSource" matSort aria-label="Elements">
+
+    <!-- Id Column -->
+    <ng-container matColumnDef="id">
+      <mat-header-cell *matHeaderCellDef mat-sort-header>Id</mat-header-cell>
+      <mat-cell *matCellDef="let row">{{row.id}}</mat-cell>
```
Comment (andrewseguin) [inline schematics/table/files/__path__/__name@dasherize@if-flat__/__name@dasherize__.component.html]: Can we make it easier for users to know what can be renamed (e.g. `row`)? Perhaps calling the data `element` instead of `row`?
